sep/sep-021.rst

```diff
@@ -46,62 +58,258 @@ Non-goals:
+Python object up the developer. Examples:
 
-Add-ons are Python modules that implement the following callbacks.
+* for a small pipeline, the add-on interface could be implemented in the same
+  class that also implements the ``open/close_spider`` and ``process_item``
+  callbacks
+* for larger add-ons, or for clearer structure, the interface could be provided
+  by a stand-alone module
 
-addon_configure
----------------
+The absolute minimum interface consists of two attributes and two callbacks:
 
-Receives the Settings object and modifies it to enable the required components.
-If it raises an exception, Scrapy will print it and exit.
+* ``NAME``: string with human-readable add-on name
+* ``VERSION``: tuple containing major/minor/patchlevel version of the add-on
+* ``update_settings()``: sets configuration (such as default values for this
+  add-on and required settings for other extensions) and enables needed
+  components.
+* ``check_configuration()``: receives the fully-initialized ``Crawler``
+  instance before it starts running, performs additional dependency and
+  configuration requirement checks
 
-Examples::
+Additionally, it may (and should, where appropriate) provide one or more
+variables that can be used for automated detection of possible dependency
+clashes:
 
-    def addon_configure(settings):
-        settings.overrides['DOWNLADER_MIDDLEWARES'].update({
-            'scrapy.contrib.downloadermiddleware.httpcache.HttpCacheMiddleware': 900,
-        })
+* ``REQUIRES``: list of built-in or custom components needed by this add-on, as
+  strings
+* ``MODIFIES``: list of built-in components whose functionality is affected or
+  replaced by this add-on (a custom HTTP cache should list ``httpcache`` here)
+* ``PROVIDES``: list of components provided by this add-on (e.g. ``mongodb``
+  for an extension that provides generic read/write access to a MongoDB
+  database, releasing other components from having to provide their own
+  database access methods)
+
+update_settings()
+-----------------
+
+Called:
+~~~~~~~
+
+Shortly after initialisation of the ``Crawler`` object.
+
+Arguments:
+~~~~~~~~~~
+
+* ``config``: the (parsed) section belong to this add-on read from
+  ``scrapy.cfg``
+* ``settings``: the crawler's ``Settings`` instance containing all project
+  settings
+* ``addon_path``: the path in which Scrapy found this callback; this will allow
+  easily enabling components defined in the same module without having to worry
+  about where it is to be found relative to the Scrapy project
+
+Purpose:
+~~~~~~~~
+
+* Modify ``settings`` to enabled required components
+* Expose some add-on specific configuration (``config``) into the global
+  settings namespace (``settings``) if necessary
+* Raise exception if components can not be properly configured (e.g. on missing
+  dependencies); Scrapy will print this exception *and exit* (making the user
+  explicitly acknowledge that the add-on does not work by forcing her to
```

Comment:
Haha thanks for the response @nyov. The sentence goes on with "disable it" and not "exit" ;) I indeed meant the user as the subject in that sentence. Sorry about the confusion, gender-sensitive language has become a huge issue in Germany over the last years and I thought using the generic she as a male writer was the "more gender-sensitive" way in English, but after you inspired me [to read up on it](https://en.wikipedia.org/wiki/Gender-specific_and_gender-neutral_pronouns#Alternatives_to_generic_he), it's apparantly considered just as sexist, whoops :hushed: I like the singular 'they' as an alternative (offtopic: [here's a cool Stephen Fry clip on 'developing language'](https://www.youtube.com/watch?v=J7E-aoXLZGY)), but then again I guess Scrapy is not a language experiment so I think I'll change everything to plural 'users'. ;)